feat(time series chart): cursor overview with multiple series support

nalaka opened the pull request:
This PR fixes issues #476, and #484.

**Problem**: The `TimeseriesChart` component has an optional `ruler` prop that configures a ruler on the chart. It is not possible to show multiple data points when multiple time series are  used.


![](https://user-images.githubusercontent.com/10908415/68017302-b4cae300-fc96-11e9-88b3-6cd8bb4b46e9.png)

**Current implementation**: The `ruler` prop is passed through to the embedded Griff `LineChart` component. 

**Change**: Inspired by the implementation in OpInt, this PR disables the ruler in Griff's `LineChart` component and implements an internal overlay component called `CursorOverview` to display the `yLabel` and `timeLabel`.

<img width="771" alt="Screenshot 2019-11-22 at 13 00 04" src="https://user-images.githubusercontent.com/437578/69406383-04497f80-0d28-11ea-8688-3daf6cced952.png">

Example usage:

```typescript
import { ChartRulerConfig, ChartRulerPoint} from '@cognite/gearbox';

<TimeseriesChart
    timeseriesIds={[123, 456]}
    startTime={Date.now() - 60 * 1000}
    endTime={Date.now()}
    ruler={{
        visible: true,
        yLabel: (point: ChartRulerPoint) =>
            `${Number(point.value).toFixed(3)}`,
        timeLabel: (point: ChartRulerPoint) =>
            new Date(point.timestamp).toISOString(),
    }}
/>
```

Where:

- `timeLabel` is an optional callback for formatting the ruler timestamp - this get called only once even with multiple points under the ruler
- `yLabel` is an optional callback for formatting the "value of each ruler point" - this gets called once for each point under the ruler

The type of the `ruler` prop is `ChartRulerConfig` - this interface is compatible with what was used previously:

```typescript
interface ChartRulerConfig {
  visible?: boolean;
  timeLabel?: (point: ChartRulerPoint) => string;
  yLabel?: (point: ChartRulerPoint) => string;
  getTimeLabelPosition?: (
    defaultPosition: number,
    measurements: {
      height: number;
      labelHeight: number;
      timeLabelMargin: number;
    }
  ) => number;
}

interface ChartRulerPoint {
  id: number | string;
  name: string;
  value: number | string;
  color: string;
  timestamp: number;
  x: number;
  y: number;
}
```

Giving an empty object for `ruler` prop uses the built-in timeLabel and yLabel callbacks.

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

polomani commented on src/components/TimeseriesChart/TimeseriesChart.tsx:
```diff
@@ -48,6 +49,7 @@ export type TimeseriesChartProps = {
 
 interface TimeseriesChartState {
   loaded: boolean;
+  legendValues: {};
```
wrong type here

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

polomani commented:
<img width="991" alt="Screenshot 2019-11-20 at 09 48 52" src="https://user-images.githubusercontent.com/10908415/69223167-22be4800-0b7b-11ea-96e4-a04fbf59c53d.png">

could we elaborate more on fake chart data here? all three timeseries are overlapping almost
could we at least apply `y-5` or something to shift some of them?

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

polomani commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+
+  render() {
+    const { series, hiddenSeries, legendValues, ruler } = this.props;
+    if (!series || series.length <= 0 || !Object.keys(legendValues).length) {
```
`!series.length` looks like the best approach for me

nalaka replied:
Yes - fixed

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

mrtraser commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+    height: 10px;
+    border-radius: 100%;
+    margin: 0 8px 0 2px;
+    background: ${props => props.color};
```
What if color is undefined here?

nalaka replied:
Good catch! Fixed.

mrtraser commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+  overviewContainer: HTMLElement | null;
+}
+export interface CursorOverviewProps {
+  styles?: CursorOverviewStyles;
```
we're trying to keep required props before an optional inside interface

nalaka replied:
Done

mrtraser commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+    );
+
+    const timeLabel =
+      typeof ruler.timeLabel === 'function'
```
There is no sense to recheck type here. Possible options – `undefined` or defined function type. So better just check if `ruler.timeLabel` available

nalaka replied:
What if the client is not using TypeScript?

Plain JS won't catch passing something like:

```js
<TimeseriesChart
    timeseriesIds={[123]}
    startTime={Date.now() - 60 * 1000}
    endTime={Date.now()}
    ruler={{
        yLabel: 100,
        timeLabel: 200,
    }}
/>
```

polomani replied:
we assume they are passing correct type, we can't spend our time on those ugly types checks.

mrtraser commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+        : formattedDate(newestTimestamp);
+
+    const yLabelFn: (val: number) => string =
+      typeof ruler.yLabel === 'function'
```
same here

mrtraser commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+    const yLabelFn: (val: number) => string =
+      typeof ruler.yLabel === 'function'
+        ? ruler.yLabel
+        : (val: number) => {
```
better to define function outside component, like `formattedDate` for exmpl

nalaka replied:
It is inside the component because the function is defined based on the `ruler` prop.

mrtraser commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+        >
+          {series && [
+            series.map(
+              (s: any) =>
```
Better to define a separate function to not mess up template

mrtraser commented on src/components/TimeseriesChart/TimeseriesChart.tsx:
```diff
@@ -107,6 +127,19 @@ export class TimeseriesChart extends React.Component<
+    });
+    this.setState({ legendValues });
+
+    if (typeof this.props.onMouseMove === 'function') {
```
Use destruction assigment for `onMouseMove` prop

polomani replied:
`typeof this.props.onMouseMove === 'function'` can be replaced with just `if (this.props.onMouseMove` if you have proper types

nalaka replied:
@mrtraser - "Use destruction assignment"?

@polomani - I was thinking of the component being used by a JS client which won't catch if a non-function is passed to `onMouseMove` prop.

polomani replied:
yes, but we don't spend our time on that
you will never be able to do all type checks. it's endless list
they should use TS

mrtraser replied:
> Use destruction assignment

```typescript
const { onMouseMove } = this.props
```

nalaka replied:
Used destruction assignment.

It is either:
```js
const { onMouseMove } = this.props;
if (typeof onMouseMove === 'function') {
      onMouseMove(data);
}
```
Or
```js
const { onMouseMove } = this.props;
if (onMouseMove) {
      onMouseMove(data);
}
```

Since we have a check anyway, going with the first version.

mrtraser commented on src/components/TimeseriesChart/TimeseriesChart.tsx:
```diff
@@ -107,6 +127,19 @@ export class TimeseriesChart extends React.Component<
     }
   };
 
+  onMouseMove = (data: any) => {
```
Can we define type here?

nalaka replied:
Currently the `onMouseMove` prop of `TimeseriesChart` has the type `(e: any) => void` - so used the same here for backward compatibility.

mrtraser commented on src/components/TimeseriesChart/TimeseriesChart.tsx:
```diff
@@ -203,11 +246,11 @@ export class TimeseriesChart extends React.Component<
                 height={height}
                 width={width}
-                ruler={ruler}
+                ruler={{ visible: false }}
```
Guess we can just remove ruler prop here

nalaka replied:
Yes - don't know what I was thinking :)

nalaka replied:
It is a required property on `LineChart` - so we need it here.

nalaka commented on src/components/AssetTimeseriesPanel/components/__snapshots__/TimeseriesPanelPure.test.tsx.snap:
```diff
@@ -2,14 +2,14 @@
 exports[`TimeseriesPanelPure Should match the snapshot 1`] = `
 <div
-  className="sc-jTzLTM jgohNR"
+  className="sc-kAzzGY iCBsrI"
```
Was generated with `jest --updateSnapshot` and the tests all pass after this.

nalaka commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+    window.removeEventListener('mousemove', this.handleMouseMove);
+  };
+
+  handleMouseMove = (e: { clientX: number; clientY: number }) => {
```
Fixed

nalaka commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+export interface CursorOverviewProps {
+  styles?: CursorOverviewStyles;
+  legendValues: LegendValues;
+  series: any;
```
`series` is defined as  `any` in `TimeseriesChartProps` as well

nalaka commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    const newestTimestamp = Object.keys(legendValues).reduce(
+      (acc: any, id: any) => {
```
There is a wierd error when I use `number` for the type. I checked on other places in Gearbox where `reduce` is used and it was used without any type annotations at all. So did that. Now no errors.

nalaka commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -0,0 +1,178 @@
+        >
+          {series && [
+            series.map(
+              (s: any) =>
```
`series` was used with `any` type throughout - maybe fix all of that later?

nalaka commented:
Pushed a commit with the fixes and suggestions incorporated:

- Fixed improper types
- Fixed unclear conditional
- Backward compatible signature for ruler callbacks `timeLabel`, `yLabel`
- Default tag color
- Moved optional props after required ones in `CursorOverviewProps`
- Documented the `RulerConfig` and `Point` values exported by `TimeseriesChart`

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

polomani commented on src/components/TimeseriesChart/stories/ruler.md:
```diff
@@ -13,14 +15,14 @@ import { TimeseriesChart } from '@cognite/gearbox';
-        yLabel: (point) =>
-          `${Number.parseFloat(point.value).toFixed(3)}`,
-        timeLabel: (point) => point.timestamp,
+        yLabel: (value: number) => Number(value).toFixed(3),
```
so is it `number` or `Point` after all?

polomani commented on src/components/TimeseriesChart/TimeseriesChart.tsx:
```diff
@@ -33,7 +33,7 @@ export type TimeseriesChartProps = {
   hiddenSeries: { [id: number]: boolean };
   annotations: Annotation[];
-  ruler: Ruler;
+  ruler: RulerConfig;
```
here is the type from griff react
```
  export interface Ruler {
    visible?: boolean;
    timeLabel?: (point: GetAggregateDatapoint) => number;
    yLabel?: (point: GetAggregateDatapoint) => string;
    timestamp?: number;
    getTimeLabelPosition?: (defaultPosition: number, params: any) => number;
  }

we need to support the same
```

polomani commented on src/components/TimeseriesChart/TimeseriesChart.tsx:
```diff
@@ -46,13 +46,37 @@ export type TimeseriesChartProps = {
+  [key: string]: LegendValue;
+}
+
+export interface Point {
```
Please use the correct type from sdk, which is `GetAggregateDatapoint`

polomani commented on src/components/TimeseriesChart/TimeseriesChart.tsx:
```diff
@@ -46,13 +46,37 @@ export type TimeseriesChartProps = {
+
+export interface RulerConfig {
+  timeLabel?: (point: Point) => string;
+  yLabel?: (point: Point) => string;
```
have a look in griff-react
type used to be 
```
  export interface Ruler {
    visible?: boolean;
    timeLabel?: (point: GetAggregateDatapoint) => number;
    yLabel?: (point: GetAggregateDatapoint) => string;
    timestamp?: number;
    getTimeLabelPosition?: (defaultPosition: number, params: any) => number;
  }

```
you are making breaking changes here

nalaka replied:
I've been working on this all morning - the type usage is a little inconsistent here, would like some help to continue.

Let's assume `Ruler` from `griff-react` is the correct type for `ruler` prop.

```typescript
export interface Ruler {
  visible?: boolean;
  timeLabel?: (point: GetAggregateDatapoint) => number;
  yLabel?: (point: GetAggregateDatapoint) => string;
  timestamp?: number;
  getTimeLabelPosition?: (defaultPosition: number, params: any) => number;
}
```

The only story where the `ruler` prop was used, had used it as below:

```typescript
ruler={{
  visible: true,
  yLabel: (point: any) =>
    `${Number.parseFloat(point.value).toFixed(3)}`,
  timeLabel: (point: any) => point.timestamp.toString(),
}}
```

Our assumption breaks down because above example's `yLabel` callback returns a `string` instead of a `number` as declared by the `Ruler` interface.

Also the `Ruler` interface uses `GetAggregateDatapoint` type in the timeLabel and yLabel callbacks. This is from `@cognite/sdk`

```typescript
interface GetAggregateDatapoint extends GetDatapointMetadata {
    average?: number;
    max?: number;
    min?: number;
    count?: number;
    sum?: number;
    interpolation?: number;
    stepInterpolation?: number;
    continuousVariance?: number;
    discreteVariance?: number;
    totalVariation?: number;
}

interface GetDatapointMetadata {
  timestamp: Date;
}
```

`GetAggregateDatapoint` does not have a `value` field, but the example uses `point.value`.

When I digged more in `griff-react` found below two types which actually match more closely - but they are not exported from `griff-react` or from `@cognite/sdk`

```typescript
export interface RulerConfig {
  visible: boolean;
  timeLabel: (data: RulerPoint) => string;
  yLabel: (data: RulerPoint) => string;
  getTimeLabelPosition?: (
    defaultPosition: number,
    measurements: {
      height: number;
      labelHeight: number;
      timeLabelMargin: number;
    }
  ) => number;
}

export interface RulerPoint {
  id?: ItemId;
  name: string;
  color: string;
  timestamp: number;
  value: number | string;
  x: number;
  y: number;
}
```

@polomani  / @mrtraser - need help on this please.

mrtraser replied:
The problem here is that `griff-react` doesn't provide any typings for TS. We're handling all related types by ourselves in gearbox based on internal implementation in `griff-react`. As for me, we need to:

* Fix `Ruler` interface

  ```typescript
  export interface Ruler {
    visible?: boolean;
    timeLabel?: (point: GetAggregateDatapoint) => string;
    yLabel?: (point: GetAggregateDatapoint) => string;
    timestamp?: number;
    getTimeLabelPosition?: (defaultPosition: number, params: any) => number;
  }
  ```
* Define type for `RulerConfig` same as in griff-react (just fixed) with omitted `timestamp` and `getTimeLabelPosition` props (cause we don't support them in any case)
* You don't need `RulerPoint` here at all cause you already have a `LegendValue` (witch pretty the same as `RulerPoint` above). So you need to use one type – `RulerPoint` or `LegendValue`. `RulerPoint` sounds better for me, but I would rename it to `ChartRulerPoint`

polomani replied:
Regarding storybook example:
```
ruler={{
  visible: true,
  yLabel: (point: any) =>
    `${Number.parseFloat(point.value).toFixed(3)}`,
  timeLabel: (point: any) => point.timestamp.toString(),
}}
```
`any` here `yLabel: (point: any) =>` and here ` timeLabel: (point: any)` actually hides the bug that we have a wrong typing in gearbox. It has to be 

```
export type DatapointsGetDatapoint =
  | DatapointsGetStringDatapoint
  | DatapointsGetDoubleDatapoint;
```
most likely (but you need to check it)

polomani replied:
@mrtraser could you help to figure out the type here? i should dive that deep into the code

nalaka replied:
@polomani / @mrtraser - So I would go with `ChartRulerPoint`.

polomani replied:
sorry, didn't see previous message

mrtraser replied:
> @polomani / @mrtraser - So I would go with `ChartRulerPoint`.

yes, please

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

polomani commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -159,17 +165,7 @@ export class CursorOverview extends React.Component<
-                )
-            ),
-          ]}
+          {series && [series.map(renderTag)]}
```
why do you have array inside an array?

nalaka replied:
It's a mistake- fixed.

polomani commented on src/components/TimeseriesChart/CursorOverview.tsx:
```diff
@@ -132,25 +132,31 @@ export class CursorOverview extends React.Component<
+          return numeral(point.value).format('0[.]0[00] a');
+        };
+
+    const renderTag = (s: any) =>
```
please use better variable name and destruct object parameter

nalaka replied:
Fixed with `const renderTag = ( {id, color}: {id: number, color: string} ) =>`

polomani commented on src/components/TimeseriesChart/TimeseriesChart.tsx:
```diff
@@ -107,6 +125,20 @@ export class TimeseriesChart extends React.Component<
+    });
+    this.setState({ rulerPoints });
+
+    if (typeof onMouseMove === 'function') {
```
still need to address this

nalaka replied:
Changed to just `if (onMouseMove) { `

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

polomani commented:
added a separate task for that

mrtraser commented:
BTW, @nalaka, can we have test here to cover basic functionality (that something renders if we provide `ruler` prop for `TimeseriesChart` component)?

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

polomani commented on src/components/TimeseriesChart/TimeseriesChart.test.tsx:
```diff
@@ -104,3 +104,23 @@ describe('TimeseriesChart', () => {
+  );
+  await sleep(300);
+  wrapper.update();
+  expect(wrapper.find('.line').exists()).toBeTruthy();
```
`expect(wrapper.find('.line').exists()).toBeTruthy();`
do you think this test is efficient enough? don't you need to look for cursor-overview instead?
you need to check if custom yLabel and timeLabel where called as well

mrtraser replied:
Can we check `CursorOverview` appearing here? (instead of success component rendering)?

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

mrtraser commented on src/components/TimeseriesChart/TimeseriesChart.test.tsx:
```diff
@@ -104,3 +105,37 @@ describe('TimeseriesChart', () => {
+  expect(wrapper.find(CursorOverview).exists()).toBeTruthy();
+});
+
+it('cursor overview does not render when ruler is not specified', async () => {
```
this test is redundant for my mind

nalaka replied:
This negative test is intended to cover the code that removes the `CursorOverview` when `ruler` prop is absent.

mrtraser replied:
I replaced it with one row in init test

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

polomani commented on src/components/TimeseriesChart/TimeseriesChart.test.tsx:
```diff
@@ -123,19 +140,16 @@ it('cursor overview renders with ruler specified', async () => {
   wrapper.update();
-  expect(wrapper.find(CursorOverview).exists()).toBeFalsy();
+
+  eventMap.mousemove({ clientX: 100, clientY: 100 });
```
cool

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

cognite-cicd commented:
This PR's storybook can now be previewed at the following URL: https://pr-501.gearboxjs.preview.cogniteapp.com

polomani commented:
closed in favour of https://github.com/cognitedata/gearbox.js/pull/535

nalaka commented:
@rumesh - only `series: any` prop in `CursorOverview.CursorOverviewProps` need to be specified a type. There is already a definition for `Series` in `src/@types/griff-react` - you will need to update it with the correct full definition. Tagging @polomani
